adapters/core2p2p/transaction.go

```diff
@@ -0,0 +1,165 @@
+					Sender:    AdaptAddress(tx.SenderAddress),
+					MaxFee:    AdaptFelt(tx.MaxFee),
+					Signature: AdaptAccountSignature(tx.Signature()),
+					ClassHash: AdaptHash(tx.ClassHash),
+					Nonce:     AdaptFelt(tx.Nonce),
+				},
+			}
+		case tx.Version.Is(2):
+			specTx.Txn = &spec.Transaction_DeclareV2_{
+				DeclareV2: &spec.Transaction_DeclareV2{
+					Sender:            AdaptAddress(tx.SenderAddress),
+					MaxFee:            AdaptFelt(tx.MaxFee),
+					Signature:         AdaptAccountSignature(tx.Signature()),
+					ClassHash:         AdaptHash(tx.ClassHash),
+					Nonce:             AdaptFelt(tx.Nonce),
+					CompiledClassHash: AdaptFelt(tx.CompiledClassHash),
+				},
+			}
+		case tx.Version.Is(3):
+			specTx.Txn = &spec.Transaction_DeclareV3_{
+				DeclareV3: &spec.Transaction_DeclareV3{
+					Sender:            AdaptAddress(tx.SenderAddress),
+					MaxFee:            AdaptFelt(tx.MaxFee),
+					Signature:         AdaptAccountSignature(tx.Signature()),
+					ClassHash:         AdaptHash(tx.ClassHash),
+					Nonce:             AdaptFelt(tx.Nonce),
+					CompiledClassHash: AdaptFelt(tx.CompiledClassHash),
+					L1Gas:             nil,
+					L2Gas:             nil,
+					Tip:               nil,
+					Paymaster:         nil,
+					NonceDomain:       "",
+					FeeDomain:         "",
+				},
+			}
+		default:
+			panic(fmt.Errorf("unsupported Declare transaction version %s", tx.Version))
+		}
+	case *core.InvokeTransaction:
+		switch {
+		case tx.Version == nil:
+			panic("Invoke transaction has not set version field")
+		case tx.Version.Is(0):
+			specTx.Txn = &spec.Transaction_InvokeV0_{
+				InvokeV0: &spec.Transaction_InvokeV0{
+					MaxFee:             AdaptFelt(tx.MaxFee),
+					Signature:          AdaptAccountSignature(tx.Signature()),
+					Address:            AdaptAddress(tx.SenderAddress), // todo for review is it ok?
+					EntryPointSelector: AdaptFelt(tx.EntryPointSelector),
+					Calldata:           AdaptFeltSlice(tx.CallData),
+				},
+			}
+		case tx.Version.Is(1):
+			specTx.Txn = &spec.Transaction_InvokeV1_{
+				InvokeV1: &spec.Transaction_InvokeV1{
+					Sender:    AdaptAddress(tx.SenderAddress),
+					MaxFee:    AdaptFelt(tx.MaxFee),
+					Signature: AdaptAccountSignature(tx.Signature()),
+					Calldata:  AdaptFeltSlice(tx.CallData),
+				},
+			}
+		default:
+			panic(fmt.Errorf("unsupported Invoke transaction version %s", tx.Version))
+		}
+	case *core.L1HandlerTransaction:
+		specTx.Txn = adaptL1HandlerTransaction(tx)
+	}
+
+	return &specTx
+}
+
+func adaptDeployTransaction(tx *core.DeployTransaction) *spec.Transaction_Deploy_ {
+	return &spec.Transaction_Deploy_{
+		Deploy: &spec.Transaction_Deploy{
+			ClassHash:   AdaptHash(tx.ClassHash),
+			AddressSalt: AdaptFelt(tx.ContractAddressSalt),
+			Calldata:    AdaptFeltSlice(tx.ConstructorCallData),
+		},
+	}
+}
+
+func adaptL1HandlerTransaction(tx *core.L1HandlerTransaction) *spec.Transaction_L1Handler {
+	if tx.Version != nil && !tx.Version.Is(1) {
```

Comment:
shouldnt this be;

```suggestion
	if tx.Version == nil || !tx.Version.Is(1) {
```

Reply:
We can assume it is non-nil btw

Reply:
Yeah because all of the transactions from the feeder have versions